src/annotator.cpp

```diff
@@ -0,0 +1,1285 @@
+            double multiplier;
+            model->units(u)->unitAttributes(i, reference, prefix, exponent, multiplier, id);
+            if (!id.empty()) {
+                idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::UNIT, std::make_pair(model->units(u), i))));
+            }
+        }
+        if (model->units(u)->isImport() && model->units(u)->importSource() != nullptr) {
+            id = model->units(u)->importSource()->id();
+            idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::IMPORT, model->units(u)->importSource())));
+        }
+    }
+    // Components.
+    for (size_t c = 0; c < model->componentCount(); ++c) {
+        listComponentIdsAndItems(model->component(c), idList);
+    }
+    // Encapsulation.
+    id = model->encapsulationId();
+    if (!id.empty()) {
+        idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::ENCAPSULATION, model)));
+    }
+
+    return idList;
+}
+
+void Annotator::build(const ModelPtr &model)
+{
+    mPimpl->isBuilt = false;
+    removeAllIssues();
+    mPimpl->mIdList.clear();
+    mPimpl->mIdList = listIdsAndItems(model);
+
+    for (auto it = mPimpl->mIdList.begin(), end = mPimpl->mIdList.end(); it != end; it = mPimpl->mIdList.upper_bound(it->first)) {
+        auto id = it->first;
+        if (mPimpl->mIdList.count(id) > 1) {
+            auto issue = libcellml::Issue::create();
+            issue->setDescription("The id '" + id + "' occurs " + std::to_string(mPimpl->mIdList.count(id)) + " times in the model. Please use the repairNonUniqueIds() function to fix them.");
+            issue->setLevel(Issue::Level::WARNING);
+            addIssue(issue);
+        }
+    }
+
+    mPimpl->isBuilt = true;
+    mPimpl->mModel = model;
+}
+
+AnyItem Annotator::item(const std::string &id)
+{
+    AnyItem item;
+    if (!mPimpl->isBuilt) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("Please call the Annotator::build function before attempting to access items by their id.");
+        issue->setLevel(libcellml::Issue::Level::ERROR);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+    if (mPimpl->mIdList.empty()) {
+        // Setting the error message to be the same as below.
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("Could not find an item with id='" + id + "' in the model.");
+        issue->setLevel(libcellml::Issue::Level::WARNING);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+
+    // Check whether the id has been duplicated in the model.
+    if (mPimpl->mIdList.count(id) > 1) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("The id '" + id + "' occurs " + std::to_string(mPimpl->mIdList.count(id)) + " times in the model. Please use the repairNonUniqueIds() function to fix them.");
```

Comment:
I don't think we should be calling out API methods in the issue messages, right? Are we doing that anywhere else?

Reply:
Not sure if we do it elsewhere.  It just seemed like the most useful info I could give!

Reply:
Moved to docstring.